File: cc/017.cc
#include <iostream>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode * left;
    TreeNode * right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        bool flag = false;
        if (pRoot1 && pRoot2)
        {
            if (pRoot1->val == pRoot2->val) flag = this->IsSubtree(pRoot1, pRoot2);
            if (!flag) flag = this->HasSubtree(pRoot1->left, pRoot2);
            if (!flag) flag = this->HasSubtree(pRoot1->right, pRoot2);
        }
        return flag;
    }
    bool IsSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        if (!pRoot2) return true;
        if (!pRoot1) return false;
        if (pRoot1->val != pRoot2->val) return false;
        return this->IsSubtree(pRoot1->left, pRoot2->left) && this->IsSubtree(pRoot1->right, pRoot2->right);
    }
};

int main(int argc, char const *argv[])
{
    TreeNode * pRoot1 = new TreeNode(0);
    pRoot1->left = new TreeNode(1);
    pRoot1->right = new TreeNode(2);
    TreeNode * pRoot2 = new TreeNode(1);
    cout << (new Solution())->HasSubtree(pRoot1, pRoot2) << endl;
    return 0;
}

File: python/017.py
# -*- coding: utf-8 -*-
# /usr/bin/python3
# Creator: Hemajun
# Description: 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）。


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        flag = False
        if pRoot1 is not None and pRoot2 is not None:
            if pRoot1.val == pRoot2.val:
                flag = self.__IsSubtree(pRoot1, pRoot2)
            if not flag:
                flag = self.HasSubtree(pRoot1.left, pRoot2)
            if not flag:
                flag = self.HasSubtree(pRoot1.right, pRoot2)
        return flag

    def __IsSubtree(self, pRoot1, pRoot2):
        if pRoot2 is None:
            return True
        if pRoot1 is None:
            return False
        if pRoot1.val != pRoot2.val:
            return False
        return self.__IsSubtree(pRoot1.left, pRoot2.left) and self.__IsSubtree(pRoot1.right, pRoot2.right)
